File: apple/RecKit/Sources/RecKit/Localization/CoreMessages.swift
import Foundation
import ReclyCore

/// docs/07 §5: the core says things in keys, the shell says them in words. This is the whole of
/// that translation on Apple — every `step_run.last_error`, every `AuthRequiredException` message
/// and every sign-in reason goes through [CoreMessages.text].
public enum CoreMessages {

    /// One core message, as the screen shows it: a sentence, and under it the diagnostic the core
    /// attached — a response body, a parser complaint — which is never translated.
    public struct Text: Equatable, Sendable {
        public let sentence: String
        public let detail: String?
    }

    /// docs/07 §5 compatibility: a `last_error` an older build wrote is a sentence, not a key, and
    /// is shown exactly as it was stored rather than replaced with a guess.
    public static func text(_ code: String) -> Text {
        guard let ref = CoreMessageRef.companion.parse(code: code) else {
            return Text(sentence: code, detail: nil)
        }
        return Text(sentence: sentence(ref.message, arg: ref.arg), detail: ref.detail)
    }

    public static func sentence(_ message: CoreMessage, arg: String? = nil) -> String {
        let key = key(for: message)
        guard takesArgument(message) else { return RecKitStrings.localized(key) }
        // The argument of this one key is itself a code — the failure that spent the last attempt —
        // so it is translated in turn and nested inside this sentence.
        let inner = message == .retryBudgetSpent ? text(arg ?? "").sentence : (arg ?? "")
        return RecKitStrings.localized(key, inner)
    }

    /// Exhaustive by construction: a new key does not compile until it has a sentence.
    static func key(for message: CoreMessage) -> String {
        switch message {
        case .needsAuth: return "Sign in again to carry on"
        case .driveReauth: return "Google Drive access has to be allowed again"
        case .driveConsentRequired: return "Drive access needs your consent"
        case .driveStorageFull: return "Google Drive is out of space — free some up and try again"
        case .signInCancelled: return "The sign-in was cancelled"
        case .missingSecret: return "This device has no value for the secret ‘%@’"
        case .invalidSecret: return "The value stored for the secret ‘%@’ is not a usable key"
        case .webhookHttp: return "The webhook answered HTTP %@"
        case .folderTemplate: return "Folder template: %@"
        case .retryBudgetSpent: return "Out of retries: %@"
        case .noRunner: return "This app cannot run a ‘%@’ step"
        case .stepMissing: return "This job’s workflow has no step ‘%@’"
        case .unsupportedStep: return "This job needs a newer version of the app: it uses a ‘%@’ step"
        case .stepFailed: return "Failed: %@"
        // docs/08 "오류": what to do about it is the whole sentence, and the provider's own line is
        // the code's detail — shown under it, never inside it.
        case .authRejected: return "The provider rejected the key."
        case .quota: return "The provider is out of quota or is rate-limiting. It will try again."
        case .providerError: return "Something went wrong at the provider. It will try again."
        case .unsupportedAudio: return "The provider would not accept this audio."
        case .noInputTrack: return "This recording has no mono or mix track to transcribe."
        case .resultTimeout: return "The provider did not finish in time. It will submit again."
        case .stale: return "The document changed while this was open"
        case .unsupportedSchema: return "Unsupported schema %@"
        }
    }

    /// The keys whose sentence has a `%@` in it; the rest are looked up without one.
    static func takesArgument(_ message: CoreMessage) -> Bool {
        switch message {
        case .needsAuth, .driveReauth, .driveConsentRequired, .driveStorageFull, .signInCancelled,
             .stale, .authRejected, .quota, .providerError, .unsupportedAudio,
             .noInputTrack, .resultTimeout:
            return false
        default:
            return true
        }
    }
}

File: windows/app/src/main/kotlin/app/recly/windows/i18n/CoreMessages.kt
package app.recly.windows.i18n

import recly.core.message.CoreMessage
import recly.core.message.CoreMessageRef

/**
 * docs/07 §5: the core hands the shell a key, the shell owns the sentence. This is the whole of
 * that translation on the desktop — every `step_run.last_error`, every `AuthRequiredException`
 * message and every sync failure reason goes through [coreMessage]. The phone's `CoreMessages`,
 * with a [Str] where its resource id is.
 */
object CoreMessages {

    /** Exhaustive by construction: a new key does not compile until it has a sentence. */
    fun keyOf(message: CoreMessage): Str = when (message) {
        CoreMessage.NEEDS_AUTH -> Str.CORE_NEEDS_AUTH
        CoreMessage.DRIVE_REAUTH -> Str.CORE_DRIVE_REAUTH
        CoreMessage.DRIVE_CONSENT_REQUIRED -> Str.CORE_DRIVE_CONSENT_REQUIRED
        CoreMessage.DRIVE_STORAGE_FULL -> Str.CORE_DRIVE_STORAGE_FULL
        CoreMessage.SIGN_IN_CANCELLED -> Str.CORE_SIGN_IN_CANCELLED
        CoreMessage.MISSING_SECRET -> Str.CORE_MISSING_SECRET
        CoreMessage.INVALID_SECRET -> Str.CORE_INVALID_SECRET
        CoreMessage.WEBHOOK_HTTP -> Str.CORE_WEBHOOK_HTTP
        CoreMessage.FOLDER_TEMPLATE -> Str.CORE_FOLDER_TEMPLATE
        CoreMessage.RETRY_BUDGET_SPENT -> Str.CORE_RETRY_BUDGET_SPENT
        CoreMessage.NO_RUNNER -> Str.CORE_NO_RUNNER
        CoreMessage.STEP_MISSING -> Str.CORE_STEP_MISSING
        CoreMessage.UNSUPPORTED_STEP -> Str.CORE_UNSUPPORTED_STEP
        CoreMessage.STEP_FAILED -> Str.CORE_STEP_FAILED
        CoreMessage.AUTH_REJECTED -> Str.CORE_AUTH_REJECTED
        CoreMessage.QUOTA -> Str.CORE_QUOTA
        CoreMessage.PROVIDER_ERROR -> Str.CORE_PROVIDER_ERROR
        CoreMessage.UNSUPPORTED_AUDIO -> Str.CORE_UNSUPPORTED_AUDIO
        CoreMessage.NO_INPUT_TRACK -> Str.CORE_NO_INPUT_TRACK
        CoreMessage.RESULT_TIMEOUT -> Str.CORE_RESULT_TIMEOUT
        CoreMessage.STALE -> Str.CORE_STALE
        CoreMessage.UNSUPPORTED_SCHEMA -> Str.CORE_UNSUPPORTED_SCHEMA
    }

    /**
     * True for the keys whose sentence has a `%1$s` in it. The ones listed below are the rest —
     * their sentence stands on its own, so they are looked up without an argument.
     */
    fun takesArgument(message: CoreMessage): Boolean = when (message) {
        CoreMessage.NEEDS_AUTH,
        CoreMessage.DRIVE_REAUTH,
        CoreMessage.DRIVE_CONSENT_REQUIRED,
        CoreMessage.DRIVE_STORAGE_FULL,
        CoreMessage.SIGN_IN_CANCELLED,
        CoreMessage.STALE,
        // docs/08 "오류": what to do about it is the whole sentence, and the provider's own line is
        // the code's detail — shown under it, never inside it.
        CoreMessage.AUTH_REJECTED,
        CoreMessage.QUOTA,
        CoreMessage.PROVIDER_ERROR,
        CoreMessage.UNSUPPORTED_AUDIO,
        CoreMessage.NO_INPUT_TRACK,
        CoreMessage.RESULT_TIMEOUT,
        -> false

        else -> true
    }
}

/**
 * A code the core wrote, as a line that is still a name.
 *
 * docs/07 §5 compatibility: a `last_error` written before the keys existed is a sentence, not a
 * key, and is carried as [UiMessage.Text] so it is shown exactly as it was stored rather than
 * replaced with a guess.
 */
fun coreMessage(code: String): UiMessage {
    val ref = CoreMessageRef.parse(code) ?: return UiMessage.Text(code)
    return coreMessage(ref.message, ref.arg)
}

/**
 * The diagnostic that came with [code], if any — a webhook's response body, a parser complaint.
 * Never translated, and never part of the sentence: the window puts it under one, in monospace.
 */
fun coreMessageDetail(code: String): String? = CoreMessageRef.parse(code)?.detail

fun coreMessage(message: CoreMessage, arg: String? = null): UiMessage {
    val key = CoreMessages.keyOf(message)
    if (!CoreMessages.takesArgument(message)) return UiMessage.Res(key)
    // The argument of this one key is itself a code — the failure that spent the last attempt — so
    // it stays a name too, nested inside this one.
    val detail: Any = when {
        arg == null -> ""
        message == CoreMessage.RETRY_BUDGET_SPENT -> coreMessage(arg)
        else -> arg
    }
    return UiMessage.Res(key, listOf(detail))
}
